File: src/main/java/cn/edu/zjnu/acm/judge/submission/ShowSourceController.java
package cn.edu.zjnu.acm.judge.submission;

import cn.edu.zjnu.acm.judge.config.LanguageFactory;
import cn.edu.zjnu.acm.judge.domain.Submission;
import cn.edu.zjnu.acm.judge.exception.MessageException;
import cn.edu.zjnu.acm.judge.mapper.SubmissionMapper;
import cn.edu.zjnu.acm.judge.mapper.UserPerferenceMapper;
import cn.edu.zjnu.acm.judge.service.SubmissionService;
import cn.edu.zjnu.acm.judge.service.UserDetailService;
import cn.edu.zjnu.acm.judge.util.ResultType;
import javax.servlet.http.HttpServletRequest;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;
import org.springframework.web.bind.annotation.RequestParam;

@Controller
@Slf4j
public class ShowSourceController {

    @Autowired
    private SubmissionMapper submissionMapper;
    @Autowired
    private UserPerferenceMapper userPerferenceMapper;
    @Autowired
    private SubmissionService submissionService;
    @Autowired
    private LanguageFactory languageFactory;

    @RequestMapping(value = "/showsource", method = {RequestMethod.GET, RequestMethod.HEAD})
    public String showsource(HttpServletRequest request,
            @RequestParam("solution_id") long submissionId,
            @RequestParam(value = "style", required = false) Integer style) {
        UserDetailService.requireLoginned(request);
        Submission submission = submissionMapper.findOne(submissionId);

        if (submission == null) {
            throw new MessageException("No such solution", HttpStatus.NOT_FOUND);
        }
        String userId = UserDetailService.getCurrentUserId(request).orElse(null);
        if (!submissionService.canView(request, submission)) {
            throw new MessageException("You have no permission to view the source.", HttpStatus.FORBIDDEN);
        }
        String language = languageFactory.getLanguage(submission.getLanguage()).getName();

        if (style == null) {
            style = userPerferenceMapper.getStyle(userId);
        } else {
            userPerferenceMapper.setStyle(userId, style);
        }
        String source = submissionMapper.findSourceById(submissionId);

        request.setAttribute("submission", submission);
        if (submission.getContest() != null) {
            request.setAttribute("contestId", submission.getContest());
        }
        request.setAttribute("language", language);
        request.setAttribute("result", ResultType.getShowsourceString(submission.getScore()));
        request.setAttribute("style", style);
        request.setAttribute("source", source);
        return "submissions/source";
    }

}

File: site/src/main/kotlin/cn/edu/zjnu/acm/judge/controller/submission/ShowSourceController.kt
package cn.edu.zjnu.acm.judge.controller.submission

import cn.edu.zjnu.acm.judge.exception.BusinessCode
import cn.edu.zjnu.acm.judge.exception.BusinessException
import cn.edu.zjnu.acm.judge.mapper.SubmissionMapper
import cn.edu.zjnu.acm.judge.mapper.SubmissionDetailMapper
import cn.edu.zjnu.acm.judge.mapper.UserPreferenceMapper
import cn.edu.zjnu.acm.judge.service.ContestOnlyService
import cn.edu.zjnu.acm.judge.service.LanguageService
import cn.edu.zjnu.acm.judge.service.SubmissionService
import cn.edu.zjnu.acm.judge.util.ResultType
import org.springframework.http.MediaType.TEXT_HTML_VALUE
import org.springframework.security.access.annotation.Secured
import org.springframework.security.core.Authentication
import org.springframework.stereotype.Controller
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import javax.servlet.http.HttpServletRequest

/**
 * @author zhanhb
 */
@Controller
@RequestMapping(produces = [TEXT_HTML_VALUE])
class ShowSourceController(
        @Suppress("SpringJavaInjectionPointsAutowiringInspection") private val submissionMapper: SubmissionMapper,
        @Suppress("SpringJavaInjectionPointsAutowiringInspection") private val userPerferenceMapper: UserPreferenceMapper,
        private val submissionService: SubmissionService,
        private val submissionDetailMapper: SubmissionDetailMapper,
        private val languageService: LanguageService,
        private val contestOnlyService: ContestOnlyService
) {
    @Secured("ROLE_USER")
    @GetMapping("showsource")
    fun showSource(request: HttpServletRequest,
                   @RequestParam("solution_id") submissionId: Long,
                   @RequestParam(value = "style", required = false) style: Int?,
                   authentication: Authentication?): String {
        var style = style
        val submission = submissionMapper.findOne(submissionId)
                ?: throw BusinessException(BusinessCode.SUBMISSION_NOT_FOUND, submissionId)

        contestOnlyService.checkViewSource(request, submission)
        val userId = authentication?.name
        if (!submissionService.canView(request, submission)) {
            throw BusinessException(BusinessCode.VIEW_SOURCE_PERMISSION_DENIED, submissionId)
        }
        val language = languageService.getLanguageName(submission.language)

        if (style == null) {
            style = userPerferenceMapper.getStyle(userId!!)
        } else {
            userPerferenceMapper.setStyle(userId!!, style)
        }
        val source = submissionDetailMapper.findSourceById(submissionId)

        request.setAttribute("submission", submission)
        if (submission.contest != null) {
            request.setAttribute("contestId", submission.contest)
        }
        request.setAttribute("language", language)
        request.setAttribute("result", ResultType.getShowsourceString(submission.score))
        request.setAttribute("style", style)
        request.setAttribute("source", source)
        return "submissions/source"
    }

}
